File: sort/912. 排序数组.java
package sort;

class Solution912 {
    public int[] sortArray(int[] nums) {
        // quickSort(nums, 0, nums.length - 1);
        mergeSort(nums, 0, nums.length - 1);
        return nums;
    }

    // 快排
    public void quickSort(int[] nums, int l, int r) {
        if (l >= r)
            return;
        int x = nums[l], i = l - 1, j = r + 1;
        while (i < j) {
            while (nums[++i] <= x) {
            }
            while (nums[--j] > x) {
            }
            if (i < j) {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
        }
        quickSort(nums, l, j);
        quickSort(nums, j + 1, r);
    }

    // 归并
    public void mergeSort(int[] nums, int l, int r) {
        if (l >= r)
            return;
        int mid = l + r >> 1;
        mergeSort(nums, l, mid);
        mergeSort(nums, mid + 1, r);
        int k = 0, i = l, j = mid + 1;
        int[] temp = new int[r - l + 1];
        while (i <= mid && j <= r) {
            if (nums[i] <= nums[j])
                temp[k++] = nums[i++];
            else
                temp[k++] = nums[j++];
        }
        while (i <= mid)
            temp[k++] = nums[i++];
        while (j <= r)
            temp[k++] = nums[j++];
        for (i = l, j = 0; i < r; i++, j++) {
            nums[i] = temp[j];
        }
    }
}

File: sort/912. 排序数组.py
from typing import List


class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:

        def quick_sort(nums, l, r):
            if l >= r:
                return
            x = nums[l]; i = l - 1; j = r + 1
            while i < j :
                i += 1
                while nums[i] < x:
                    i += 1
                j -= 1
                while nums[j] > x:
                    j -= 1
                if i < j:
                    nums[i],nums[j] = nums[j],nums[i]
            quick_sort(nums, l, j)
            quick_sort(nums, j+1, r)

        quick_sort(nums, 0, len(nums)-1)
        return nums


if __name__ == "__main__":
    s = Solution()
    nums = [5,1,1,6,0,0]
    s.sortArray(nums=nums)
    print(nums)

# 2 4 6 3 6 1 0
